core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,430 @@
+            } else {
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // SurroundingSubscriber is publishing data.
+                // New requests will be handled by 'publishDownstream(item)'.
+                return;
+            }
+
+            publish();
+        }
+
+        private void publish() {
+            if (closed || requested <= 0 && upstreamRequested <= 0) {
+                return;
+            }
+
+            switch (state) {
+                case REQUIRE_HEAD: {
+                    sendHead();
+                    break;
+                }
+                case REQUIRE_BODY: {
+                    if (!subscribed) {
+                        subscribed = true;
+                        publisher.subscribe(this, executor, options);
+                        return;
+                    }
+                    if (upstreamRequested > 0) {
+                        return;
+                    }
+                    final Subscription upstream = this.upstream;
+                    if (upstream != null) {
+                       requestUpstream(upstream);
+                    }
+                    break;
+                }
+                case REQUIRE_TAIL: {
+                    sendTail();
+                    break;
+                }
+                case REQUIRE_COMPLETE: {
+                    sendComplete();
+                    break;
+                }
+                case DONE: {
+                    closed = true;
+                    break;
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            publishDownstream(head);
+        }
+
+        private void sendTail() {
+            setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
+            if (tail != null) {
+                publishDownstream(tail);
+            } else {
+                sendComplete();
+            }
+        }
+
+        private void sendComplete() {
+            setState(State.REQUIRE_COMPLETE, State.DONE);
+            close0(null);
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            if (requested <= 0) {
+                return;
+            }
+            assert upstreamRequested == 0;
+            upstreamRequested = requested;
+            if (requested < Long.MAX_VALUE) {
+                requested = 0;
+            }
+            subscription.request(upstreamRequested);
+        }
+
+        private void publishDownstream(T item) {
+            requireNonNull(item, "item");
+            if (closed) {
+                return;
+            }
+            downstream.onNext(item);
+            if (0 < upstreamRequested && upstreamRequested < Long.MAX_VALUE) {
+                upstreamRequested--;
+            } else if (requested < Long.MAX_VALUE) {
+                requested--;
```

Comment:
nit:
```suggestion
                // Have to decrease when the head is published.
                requested--;
```

Reply:
On second thoughts, how about making it clearer?
```java
private void sendHead() {
    setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
    publishDownstream(head, true);
}

private void sendTail() {
    setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
    if (tail != null) {
        downstream.onNext(tail);
    }
    sendComplete();
}

private void publishDownstream(T item, boolean head) {
    requireNonNull(item, "item");
    if (closed) {
        return;
    }
    downstream.onNext(item);
    if (head) {
        if (requested < Long.MAX_VALUE) {
            requested--;
        }
    } else {
        assert upstreamRequested > 0;
        if (upstreamRequested < Long.MAX_VALUE) {
            upstreamRequested--;
        }
    }

    if (!publishedAny) {
        publishedAny = true;
    }

    publish();
}
```

Reply:
✅ updated!

Reply:
Sorry for making you do this twice. 😢

Reply:
```
publishDownstream(head, true);
```
aha we can pass curState as param! I'll update it~!

Reply:
> I found that we change state first, and then call publishDownstream.

That's why I used `boolean head` instead of the state. 😄 
One of the reasons I asked you to change is that we use `requested` for the `head` and `upstreamRequested` for the body, but it's hard to distinguish in the method.
Also, we don't have to decrease the value when the `tail` is sent and we can call `sendComplete()` right away after sending `tail`. 
WDYT?

Reply:
✅ updated! looks much better. thanks for suggestion 👍

Reply:
> Also, we don't have to decrease the value when the tail is sent and we can call sendComplete() right away after sending tail.

oh we have to decrease `upstreamRequested` or `requested` on tail, cause **we shouldn't sendComplete right away after sending tail If `requested == 0 && upstreamRequested==0` after sendTail.**

I found that it's Reactor's publisher spec~! If we sendComplete right after sendTail on `requested == 0 && upstreamRequested==0`, TCk test failed!

Reply:
I believe we can call `sendComplete`. 🤔 
https://github.com/reactive-streams/reactive-streams-jvm#2.9
```
A Subscriber MUST be prepared to receive an onComplete signal with or without a preceding Subscription.request(long n) call.
```

Reply:
> I found that it's Reactor's publisher spec~!

Ah, you mean it is Reactor's spec. 😓 
I don't know why they violate the spec but let's keep it as is then.

Reply:
```
private void sendTail() {
    setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
    if (tail != null) {
        downstream.onNext(tail);
    }
    sendComplete(); // always sendComplete right after sendTail, although requested only one.
}
```
You can check TCK test failed with above code~! 🙇

Reply:
Could you tell me which test failed? 🤔

Reply:
oh I misunderstood some reactive-stream spec 😅 updated!